### PR TITLE
feat(optimizer): bushy join equivalent condition derivation

### PR DESCRIPTION
**Pull request by KveinAxel**
Signed-off-by: Kevin Axel <kevinaxel@163.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

support bushy join equivalent condition derivation, https://github.com/risingwavelabs/risingwave/pull/8316#pullrequestreview-1345479333

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/bushy_join.yaml [hunk 1]
```diff
@@ -604,35 +567,6 @@
     order by
       nation,
       o_year desc;
-  stream_plan: |
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/bushy_join.yaml:607]: Does the streaming plan miss?

File: src/frontend/src/optimizer/plan_node/logical_multi_join.rs [hunk 2]
```diff
@@ -727,6 +690,79 @@ impl LogicalMultiJoin {
...
+    for node in keys {
+        let rel = nodes
+            .get(&node)
+            .unwrap()
+            .relations
+            .iter()
+            .cloned()
+            .collect_vec();
+
+        let n_rel = rel.len();
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/logical_multi_join.rs:723]: After that, does `rel` never access again?

File: src/frontend/planner_test/tests/testdata/bushy_join.yaml (hunk 1 above)
Comment (KveinAxel) [reply src/frontend/planner_test/tests/testdata/bushy_join.yaml:607]: I rerun the apply-planner-test, and find
```
<   stream_error: |-
<     Not supported: streaming nested-loop join
<     HINT: The non-equal join in the query requires a nested-loop join executor, which could be very expensive to run. Consider rewriting the query to use dynamic filter as a substitute if possible.
<     See also: https://github.com/risingwavelabs/rfcs/blob/main/rfcs/0033-dynamic-filter.md
```

File: src/frontend/src/optimizer/plan_node/logical_multi_join.rs [hunk 3]
```diff
@@ -727,6 +690,79 @@ impl LogicalMultiJoin {
...
+#[derive(Clone, Debug)]
+struct GraphNode {
+    id: usize,
+    join_tree: JoinTreeNode,
+    // use BTreeSet for deterministic
+    relations: BTreeSet<usize>,
+}
+
+///  equivalent condition derivation by `a = b && a = c` ==> `b = c`
+fn eq_condition_derivation(mut nodes: BTreeMap<usize, GraphNode>) -> BTreeMap<usize, GraphNode> {
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/logical_multi_join.rs:712]: In addition to adding edges between graph nodes, we need to add the corresponding equal conditions.

File: src/frontend/planner_test/tests/testdata/bushy_join.yaml [hunk 4]
```diff
@@ -609,30 +609,31 @@
...
-                          | └─StreamProject { exprs: [part.p_partkey] }
-                          |   └─StreamFilter { predicate: Like(part.p_name, '%yellow%':Varchar) }
-                          |     └─StreamTableScan { table: part, columns: [part.p_partkey, part.p_name], pk: [part.p_partkey], dist: UpstreamHashShard(part.p_partkey) }
-                          └─StreamExchange { dist: HashShard(lineitem.l_partkey) }
-                            └─StreamTableScan { table: lineitem, columns: [lineitem.l_orderkey, lineitem.l_partkey, lineitem.l_suppkey, lineitem.l_quantity, lineitem.l_extendedprice, lineitem.l_discount, lineitem.l_linenumber], pk: [lineitem.l_orderkey, lineitem.l_linenumber], dist: UpstreamHashShard(lineitem.l_orderkey, lineitem.l_linenumber) }
+          └─StreamProject { exprs: [nation.n_name, Extract('YEAR':Varchar, orders.o_orderdate) as $expr1, ((lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount)) - (partsupp.ps_supplycost * lineitem.l_quantity)) as $expr2, partsupp.ps_partkey, partsupp.ps_suppkey, nation.n_nationkey, supplier.s_suppkey, part.p_partkey, orders.o_orderkey, lineitem.l_orderkey, lineitem.l_linenumber] }
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/bushy_join.yaml:612]: The TPCH q9 presents a slight deviation from our anticipated join tree, despite both having an equal join height of 3. It appears that the current algorithm selects `merge_node` from `for merge_node in &n.relations {` without any priority consideration. In my opinion, we should prioritize selecting `merge_node` with lower join tree heights and fewer adjacent nodes.

File: src/frontend/planner_test/tests/testdata/bushy_join.yaml [hunk 5]
```diff
@@ -609,30 +609,31 @@
...
+            └─StreamHashJoin { type: Inner, predicate: supplier.s_suppkey = lineitem.l_suppkey AND partsupp.ps_suppkey = lineitem.l_suppkey AND partsupp.ps_partkey = lineitem.l_partkey AND partsupp.ps_partkey = part.p_partkey, output: [lineitem.l_quantity, lineitem.l_extendedprice, lineitem.l_discount, partsupp.ps_supplycost, orders.o_orderdate, nation.n_name, partsupp.ps_partkey, partsupp.ps_suppkey, nation.n_nationkey, supplier.s_suppkey, part.p_partkey, orders.o_orderkey, lineitem.l_orderkey, lineitem.l_linenumber] }
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/bushy_join.yaml:613]: Let us do some pruning to reduce the join keys.

Comment (st1page) [reply src/frontend/planner_test/tests/testdata/bushy_join.yaml:613]: https://github.com/risingwavelabs/risingwave/issues/8376

File: src/frontend/planner_test/tests/testdata/bushy_join.yaml [hunk 6]
```diff
@@ -359,32 +359,32 @@
     └─StreamProject { exprs: [nation.n_name, sum($expr1)] }
       └─StreamHashAgg { group_key: [nation.n_name], aggs: [sum($expr1), count] }
         └─StreamExchange { dist: HashShard(nation.n_name) }
-          └─StreamProject { exprs: [nation.n_name, (lineitem.l_extendedprice * (1:Int32 - lineitem.l_discount)) as $expr1, lineitem.l_orderkey, lineitem.l_linenumber, customer.c_custkey, orders.o_orderkey, region.r_regionkey, nation.n_nationkey, supplier.s_suppkey, lineitem.l_suppkey, customer.c_nationkey] }
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/bushy_join.yaml:362]: We copy the whole `tpch.yaml` file to verify the bushy join reordering, but some of them are useless, could you please remove those cases? Additionally, could you provide a test case comprising of no less than 20 inner joins to validate if this present algorithm can efficiently handle large-scale join operations?

**Review by chenzl25 — APPROVED** [review]
Rest LGTM.

**Pull request merged**